packages/@uppy/companion-client/src/Provider.ts

```diff
@@ -0,0 +1,445 @@
+  B
+> {
+  files: UppyFile<M, B>[]
+
+  storage: typeof tokenStorage
+}
+
+const getName = (id: string) => {
+  return id
+    .split('-')
+    .map((s) => s.charAt(0).toUpperCase() + s.slice(1))
+    .join(' ')
+}
+
+function getOrigin() {
+  // eslint-disable-next-line no-restricted-globals
+  return location.origin
+}
+
+function getRegex(value: string | RegExp) {
+  if (typeof value === 'string') {
+    return new RegExp(`^${value}$`)
+  }
+  if (value instanceof RegExp) {
+    return value
+  }
+  return undefined
+}
+
+function isOriginAllowed(
+  origin: string,
+  allowedOrigin: string | Array<string | RegExp>,
+) {
+  const patterns = Array.isArray(allowedOrigin)
+    ? allowedOrigin.map(getRegex)
+    : [getRegex(allowedOrigin)]
+  return patterns.some(
+    (pattern) => pattern?.test(origin) || pattern?.test(`${origin}/`),
+  ) // allowing for trailing '/'
+}
+
+export default class Provider<
+  M extends Meta,
+  B extends Body,
+> extends RequestClient<M, B> {
+  #refreshingTokenPromise: Promise<void> | undefined
+
+  provider: string
+
+  id: string
+
+  name: string
+
+  pluginId: string
+
+  tokenKey: string
+
+  companionKeysParams?: Record<string, string>
+
+  preAuthToken: string | null
+
+  supportsRefreshToken: boolean
+
+  constructor(uppy: Uppy<M, B>, opts: Opts) {
+    super(uppy, opts)
+    this.provider = opts.provider
+    this.id = this.provider
+    this.name = this.opts.name || getName(this.id)
+    this.pluginId = this.opts.pluginId
+    this.tokenKey = `companion-${this.pluginId}-auth-token`
+    this.companionKeysParams = this.opts.companionKeysParams
+    this.preAuthToken = null
+    this.supportsRefreshToken = opts.supportsRefreshToken ?? true // todo false in next major
+  }
+
+  async headers(): Promise<Record<string, string>> {
+    const [headers, token] = await Promise.all([
+      super.headers(),
+      this.#getAuthToken(),
+    ])
+    const authHeaders: Record<string, string> = {}
+    if (token) {
+      authHeaders['uppy-auth-token'] = token
+    }
+
+    if (this.companionKeysParams) {
+      authHeaders['uppy-credentials-params'] = btoa(
+        JSON.stringify({ params: this.companionKeysParams }),
+      )
+    }
+    return { ...headers, ...authHeaders }
+  }
+
+  onReceiveResponse(response: Response): Response {
+    super.onReceiveResponse(response)
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return response
+    const oldAuthenticated = plugin.getPluginState().authenticated
+    const authenticated = oldAuthenticated
+      ? response.status !== authErrorStatusCode
+      : response.status < 400
+    plugin.setPluginState({ authenticated })
+    return response
+  }
+
+  async setAuthToken(token: string): Promise<void> {
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return undefined
+    // @ts-expect-error don't think we can make core aware that _some_ plugins have a storage
+    return plugin.storage.setItem(this.tokenKey, token)
```

Comment:
This method has changed a lot, and is using `@ts-expect-error` anyway, let's rollback to what it was before, no?
```suggestion
    // @ts-expect-error don't think we can make core aware that _some_ plugins have a storage
    return this.uppy.getPlugin(this.pluginId).storage.setItem(this.tokenKey, token);
```